segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/OperationProcessor.java

```diff
@@ -366,24 +365,34 @@ private boolean isFatalException(Throwable ex) {
     /**
      * Temporary State for the QueueProcessor. Keeps track of pending Operations and allows committing or failing all of them.
      */
-    @Slf4j
+    @ThreadSafe
     private static class QueueProcessingState {
         private final String traceObjectId;
-        private final Queue<CompletableOperation> pendingOperations;
+        private final Object lock;
+        @GuardedBy("lock")
+        private final Deque<CompletableOperation> pendingOperations;
         private final OperationMetadataUpdater metadataUpdater;
         private final MemoryStateUpdater logUpdater;
         private final MetadataCheckpointPolicy checkpointPolicy;
-
-        QueueProcessingState(OperationMetadataUpdater metadataUpdater, MemoryStateUpdater stateUpdater, MetadataCheckpointPolicy checkpointPolicy, String traceObjectId) {
-            assert metadataUpdater != null : "metadataUpdater is null";
-            assert stateUpdater != null : "stateUpdater is null";
-            assert checkpointPolicy != null : "checkpointPolicy is null";
-
+        private final Consumer<Throwable> fatalExceptionCallback;
+        @GuardedBy("lock")
+        private final SortedDeque<DataFrameBuilder.CommitArgs> metadataTransactions;
+        @GuardedBy("lock")
+        private long highestCommittedDataFrame;
+        @GuardedBy("lock")
+        private Throwable stopException;
+
+        QueueProcessingState(OperationMetadataUpdater metadataUpdater, MemoryStateUpdater stateUpdater, MetadataCheckpointPolicy checkpointPolicy,
+                             Consumer<Throwable> fatalExceptionCallback, Object lock, String traceObjectId) {
+            this.metadataUpdater = Preconditions.checkNotNull(metadataUpdater, "metadataUpdater");
+            this.logUpdater = Preconditions.checkNotNull(stateUpdater, "stateUpdater");
+            this.checkpointPolicy = Preconditions.checkNotNull(checkpointPolicy, "checkpointPolicy");
+            this.fatalExceptionCallback = Preconditions.checkNotNull(fatalExceptionCallback, "fatalExceptionCallback");
+            this.lock = Preconditions.checkNotNull(lock, "lock");
```

Comment:
As discussed on Slack, I made this class non-static which removed the need for this, as well as a few more fields.